### PR TITLE
Receipt chain hashes for Parties.t

### LINKED ISSUE #10350: Receipt chain hash updates for parties transactions

### PR DESCRIPTION
**Pull request by psteckler**
Update the receipt chain hashes for each account associated with a party in a `Parties.t`. The new hash is computed from the old hash, the full transaction commitment of the `Parties.t` and a 0-based index indicating the order in which the party is processed during transaction application.

Separate `Receipt.Chain_hash.cons` into two separate functions, `cons_signed_command_playload` and `cons_parties_commitment`, and separate `Elt` into two distinct types to give as inputs to those functions. Add a test that the un/checked hashes correspond for `Parties.t`.

An addition field `party_index` is added to the parties local state. The sample precomputed block is edited (once again) to add that field.

In the unchecked world, a transaction commitment had been the unit value. In order to compute receipt chain hashes, that becomes a bona fide field value.

Because accounts get updated receipt chain hashes, that changes ledger hashes. The ledger hashes are modified in the replayer test data to get the replayer test to pass.

An account associated with a `Parties.t` may have its receipt chain hash updated several times. In the parties generator, use `Ignore` for the receipt chain hash precondition for now, since the precondition will only be satisfied for the first update. See issue #11454.

The CLI command `advanced verify-receipt` fails explicitly when given zkApps (it never worked correctly for that case). See issue #11431.

Note: the version linter will fail in CI due to changes in parties-related types, which haven't been released.

Closes #10350.

### CODE REVIEW

File: src/lib/transaction_logic/mina_transaction_logic.ml [hunk 1]
```diff
@@ -975,15 +978,31 @@ module Make (L : Ledger_intf.S) : S with type ledger := L.t = struct
...
+    end
+
+    module Length = struct
+      type t = Mina_numbers.Length.t
 
-      let commitment ~other_parties:_ = ()
+      let zero, succ = Mina_numbers.Length.(zero, succ)
 
-      let full_commitment ~party:_ ~memo_hash:_ ~commitment:_ = ()
+      let if_ = Parties.value_if
```
Comment (deepthiskumar) [inline src/lib/transaction_logic/mina_transaction_logic.ml:1005]: This is a bit confusing. Maybe define the `value_if` function in this file?

File: src/lib/transaction_logic/parties_logic.ml [hunk 2]
```diff
@@ -1383,6 +1433,26 @@ module Make (Inputs : Inputs_intf) = struct
       let a = Account.set_voting_for voting_for a in
       (a, local_state)
     in
+    (* Update receipt chain hash *)
+    let a =
+      let new_hash =
+        let old_hash = Account.receipt_chain_hash a in
+        Receipt_chain_hash.if_
+          (let open Inputs.Bool in
+          Party.has_signature_authorization party
```
Comment (deepthiskumar) [inline src/lib/transaction_logic/parties_logic.ml:1442]: Checking `signature_verifies ||| proof_verifies` is sufficient here

**Review by mrmr1993 — APPROVED** [review]

File: src/lib/mina_commands/mina_commands.ml [hunk 3]
```diff
@@ -157,14 +157,20 @@ module Receipt_chain_verifier = Merkle_list_verifier.Make (struct
...
+    match proof_elem with
+    | Signed_command cmd ->
+        let elt =
+          Receipt.Signed_command_elt.Signed_command_payload
+            (Signed_command.payload cmd)
+        in
+        Receipt.Chain_hash.cons_signed_command_payload elt parent_hash
+    | Parties _parties ->
+        failwith "Not implemented for zkApps"
+  (* TODO: apply cons_parties_commitment operation for all occurrences of a party
```
[the remark this answers is not in the document]
Comment (psteckler) [reply src/lib/mina_commands/mina_commands.ml]: Added issue #11495.

There's another, related issue, #11431, for `compute-receipt-chain-hash`.

Those could both be done in a separate PR.

**Review by lk86 — APPROVED** [review]

**Review by bkase — APPROVED** [review]
Product review ✅

**Pull request merged**